cmd/k8s-operator/sts.go

```diff
@@ -763,16 +765,37 @@ func applyProxyClassToStatefulSet(pc *tsapi.ProxyClass, ss *appsv1.StatefulSet,
 	return ss
 }
 
-func enableMetrics(ss *appsv1.StatefulSet, pc *tsapi.ProxyClass) {
+func enableMetrics(ss *appsv1.StatefulSet) {
 	for i, c := range ss.Spec.Template.Spec.Containers {
 		if c.Name == "tailscale" {
-			// Serve metrics on on <pod-ip>:9001/debug/metrics. If
-			// we didn't specify Pod IP here, the proxy would, in
-			// some cases, also listen to its Tailscale IP- we don't
-			// want folks to start relying on this side-effect as a
-			// feature.
-			ss.Spec.Template.Spec.Containers[i].Env = append(ss.Spec.Template.Spec.Containers[i].Env, corev1.EnvVar{Name: "TS_TAILSCALED_EXTRA_ARGS", Value: "--debug=$(POD_IP):9001"})
-			ss.Spec.Template.Spec.Containers[i].Ports = append(ss.Spec.Template.Spec.Containers[i].Ports, corev1.ContainerPort{Name: "metrics", Protocol: "TCP", HostPort: 9001, ContainerPort: 9001})
+			ss.Spec.Template.Spec.Containers[i].Env = append(ss.Spec.Template.Spec.Containers[i].Env,
+				// Serve tailscaled's debug metrics on on
+				// <pod-ip>:9001/debug/metrics. If we didn't specify Pod IP
+				// here, the proxy would, in some cases, also listen to its
+				// Tailscale IP- we don't want folks to start relying on this
+				// side-effect as a feature.
+				corev1.EnvVar{
+					Name:  "TS_TAILSCALED_EXTRA_ARGS",
+					Value: "--debug=$(POD_IP):9001",
+				},
+				// Serve client usermetrics on <pod-ip>:9002/metrics.
+				corev1.EnvVar{
+					Name:  "TS_USERMETRICS_ADDR_PORT",
+					Value: "$(POD_IP):9002",
+				},
+			)
+			ss.Spec.Template.Spec.Containers[i].Ports = append(ss.Spec.Template.Spec.Containers[i].Ports,
+				corev1.ContainerPort{
+					Name:          "metrics",
+					Protocol:      "TCP",
+					ContainerPort: 9001,
+				},
+				corev1.ContainerPort{
+					Name:          "usermetrics",
+					Protocol:      "TCP",
+					ContainerPort: 9002,
+				},
+			)
```

Comment:
The naming of the `metrics` port was unfortunate. Because this is also the point of integration for users right now, and because I would like it make the distinction between 'stable' and 'debug' metrics more clear, I'm thinking that perhaps we could serve the new user metrics on `metrics` port and rename the other port to `debugmetrics` or similar? This would be a breaking change for anyone who has scripted the name, but we did document that these metrics will likely change in breaking ways on ProxyClass

Reply:
If the existing `metrics` port exposes the tailscaled debug web server, perhaps we could call it `debug` port (w/o metrics in the name)? That web server already serves pprof endpoints in addition to metrics, and we might add more in the future: https://github.com/tailscale/tailscale/blob/b9ecc50ce38d23cfacd5fae6360fa9742c0564a6/cmd/tailscaled/tailscaled.go#L763-L772

Reply:
That would make sense if we are renaming it anyway 👍🏼